### PR TITLE
Notes about unique vs cat.categories #12278

### LINKED ISSUE #12278: Performance: .unique / .nunique of categorical series slow on large data set
I've noticed that `Series.unique` and `Series.nunique` when used with a categorical series can be slow on large dataset. Presumably its not utilising the shortcuts:
```
Series.unique = Series.cat.categories
Series.nunique = len(Series.cat.categories)
``` 

Heres an example in iPhython:
```
s = pd.Series(np.random.choice(['a','b','c'], 100000000)).astype('category')
%time s.nunique()
896 ms
%time len(s.cat.categories)
55.1 µs
```

Its significantly slower indicating its not using the above shortcut.

pd.show_versions()

```
pandas: 0.17.1
Cython: 0.23.4
numpy: 1.10.1
IPython: 4.0.1
```

### PR DESCRIPTION
**Pull request by Dorozhko-Anton**
- [x] closes #12278
- [ ] tests added / passed
- [x] passes `git diff upstream/master | flake8 --diff`
- [x] adding notes about .unique() and .cat.categories (as suggested in [#12278 (comment)](https://github.com/pydata/pandas/issues/12278#issuecomment-182445522))

### CODE REVIEW

File: doc/source/categorical.rst [hunk 1]
```diff
@@ -182,6 +182,19 @@ It's also possible to pass in the categories in a specific order:
     indicate an ordered ``Categorical``.
 
 
+.. note::
+
+    Pay attention! Result of ``.unique()`` not always the same as ``Series.cat.categories``,
```
Comment (TomAugspurger) [inline doc/source/categorical.rst]: Don't need the "Pay attention". Add a "The" before "result". Let's change ``.unique()`` to ``Series.unique()`` to make that a bit clearer.

Comment (jreback) [inline doc/source/categorical.rst]: dont' say 'Pay attention!',  you can do a ``::warning::`` if you want, but a note is just fine.

``The results of .unique() are not always the same as .categories.....``

File: doc/source/categorical.rst [hunk 2]
```diff
@@ -182,6 +182,19 @@ It's also possible to pass in the categories in a specific order:
...
+    because ``.unique()`` has a couple of guarantees, namely that it returns categories in the order of appearance,
+    and it only includes values that are actually present.
+    See example below:
+
+.. ipython:: python
```
Comment (jreback) [inline doc/source/categorical.rst]: indent so this is in the note

File: doc/source/categorical.rst [hunk 3]
```diff
@@ -182,6 +182,19 @@ It's also possible to pass in the categories in a specific order:
     indicate an ordered ``Categorical``.
 
 
+.. note::
+
+    Pay attention! Result of ``.unique()`` not always the same as ``Series.cat.categories``,
+    because ``.unique()`` has a couple of guarantees, namely that it returns categories in the order of appearance,
+    and it only includes values that are actually present.
+    See example below:
```
Comment (jreback) [inline doc/source/categorical.rst]: you don't need see example below

File: doc/source/categorical.rst [hunk 4]
```diff
@@ -182,6 +182,19 @@ It's also possible to pass in the categories in a specific order:
...
+
+.. ipython:: python
+
+     s = pd.Series(list('babc')).astype('category', categories=list('abcd'))
+     s.unique()
```
Comment (jreback) [inline doc/source/categorical.rst]: show s

File: pandas/core/categorical.py [hunk 5]
```diff
@@ -118,6 +118,10 @@ def maybe_to_categorical(array):
 
 Assigning to `categories` is a inplace operation!
 
+Note
+----
+Categories getter returns all categories and not only those present in the data
```
Comment (jreback) [inline pandas/core/categorical.py]: this doesn't below here. not really sure of the point you are trying to make

File: pandas/core/categorical.py [hunk 6]
```diff
@@ -118,6 +118,10 @@ def maybe_to_categorical(array):
 
 Assigning to `categories` is a inplace operation!
 
+Note
```
Comment (TomAugspurger) [inline pandas/core/categorical.py]: This is supposed to be `Notes` (with an `s`). I'm not sure that the section is actually needed though.

File: pandas/core/categorical.py (hunk 5 above)
Comment (Dorozhko-Anton) [inline pandas/core/categorical.py]: My point was to implement first bulet in [#12278 comment](https://github.com/pydata/pandas/issues/12278#issuecomment-182445522)
- [ ] add a note to the docstring of categories that this includes all categories and not only those present in the data

But I agree that it is a reduntant note

File: doc/source/categorical.rst (hunk 1 above)
Comment (jreback) [inline doc/source/categorical.rst]: @TomAugspurger hah, I think we say it at the same time!

File: doc/source/categorical.rst [hunk 7]
```diff
@@ -182,6 +182,19 @@ It's also possible to pass in the categories in a specific order:
     indicate an ordered ``Categorical``.
 
 
+.. note::
+
+    The result of ``Series.unique()`` not always the same as ``Series.cat.categories``,
```
Comment (jorisvandenbossche) [inline doc/source/categorical.rst]: "*is* not always .."

File: pandas/core/categorical.py (hunk 5 above)
Comment (Dorozhko-Anton) [inline pandas/core/categorical.py]: Should I rebase and force push to delete the commit that alters categories docstring?

Comment (jorisvandenbossche) [inline pandas/core/categorical.py]: @Dorozhko-Anton current state looks good! Can you just squash the commits into one?

Comment (Dorozhko-Anton) [inline pandas/core/categorical.py]: @jorisvandenbossche sure.

**Comment by jreback** [comment]
thanks @Dorozhko-Anton

**Pull request closed without merging**